azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -44,10 +51,79 @@ def __init__(self, callback):
                 "Cannot instantiate PipelineOperation object.  You need to use a derived class"
             )
         self.name = self.__class__.__name__
-        self.callback = callback
+        self.callbacks = []
         self.needs_connection = False
         self.completed = False
 
+        self.callbacks.append(callback)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def add_callback(self, callback):
+        self.callbacks.append(callback)
```

Comment:
>append [](start = 23, length = 6)

Something to think about:  If our collection of callbacks is a list (in the abstract sense), then the call order is not obvious (do we call the first item in the list first, or do we call the last item first?).  Since we're using append here, i think we call the last one first, but that's the opposite of what I would assume.  

However, if we think of our collection of callbacks as a stack, then the order is much more intuitive.  When we add a callback, we push it on the stack.  When the operation is complete, we pop the first callback and call it, then we pop the next.  This matches the order we want for callbacks (LIFO).

Reply:
Agree, one of the internal changes I've been considering is to refer to this as the `callback_stack`

---
In reply to: [345985415](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r345985415) [](ancestors = 345985415)